src/Compilers/Core/CodeAnalysisTest/Diagnostics/SystemLanguageAnalyzer.cs

```diff
@@ -128,4 +132,118 @@ void Report(CompilationAnalysisContext context, IFieldSymbol field, DiagnosticDe
+    public class LocalCouldBeConstAnalyzer : DiagnosticAnalyzer
+    {
+        private const string SystemCategory = "System";
+
+        public static readonly DiagnosticDescriptor LocalCouldBeConstDescriptor = new DiagnosticDescriptor(
+            "LocalCouldBeReadOnly",
+            "Local Could Be Const",
+            "Local variable is never modified and so could be const.",
+            SystemCategory,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        /// <summary>Gets the set of supported diagnostic descriptors from this analyzer.</summary>
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get { return ImmutableArray.Create(LocalCouldBeConstDescriptor); }
+        }
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.RegisterOperationBlockStartAction(
+                (operationBlockContext) =>
+                {
+                    IMethodSymbol containingMethod = operationBlockContext.OwningSymbol as IMethodSymbol;
+
+                    if (containingMethod != null)
+                    {
+                        HashSet<ILocalSymbol> mightBecomeConstLocals = new HashSet<ILocalSymbol>();
+                        HashSet<ILocalSymbol> assignedToLocals = new HashSet<ILocalSymbol>();
+
+                        operationBlockContext.RegisterOperationAction(
+                           (operationContext) =>
+                           {
+                               IAssignmentExpression assignment = (IAssignmentExpression)operationContext.Operation;
+                               AssignTo(assignment.Target, assignedToLocals, mightBecomeConstLocals);
+                           },
+                           OperationKind.AssignmentExpression,
+                           OperationKind.CompoundAssignmentExpression,
+                           OperationKind.IncrementExpression);
+
+                        operationBlockContext.RegisterOperationAction(
+                            (operationContext) =>
+                            {
+                                IInvocationExpression invocation = (IInvocationExpression)operationContext.Operation;
+                                foreach (IArgument argument in invocation.ArgumentsInParameterOrder)
+                                {
+                                    if (argument.Parameter.RefKind == RefKind.Out || argument.Parameter.RefKind == RefKind.Ref)
+                                    {
+                                        AssignTo(argument.Value, assignedToLocals, mightBecomeConstLocals);
+                                    }
+                                }
+                            },
+                            OperationKind.InvocationExpression);
+
+                        operationBlockContext.RegisterOperationAction(
```

Comment:
If only we invoked SymbolAnalyzers for LocalSymbols then this could be simpler..